python/xorbits/deploy/kubernetes/client.py

```diff
@@ -0,0 +1,506 @@
+
+    def _load_cluster_logs(self) -> Dict:  # pragma: no cover
+        log_dict = dict()
+        pod_items = self._core_api.list_namespaced_pod(self._namespace).to_dict()
+        for item in pod_items["items"]:
+            log_dict[item["metadata"]["name"]] = self._core_api.read_namespaced_pod_log(
+                name=item["metadata"]["name"], namespace=self._namespace
+            )
+        return log_dict
+
+    def start(self):
+        try:
+            self._create_namespace()
+            self._create_roles_and_bindings()
+
+            self._create_services()
+            self._create_kube_service()
+
+            self._wait_services_ready()
+
+            self._create_ingress()
+
+            self._external_web_endpoint = self._get_ingress_address()
+            self._wait_web_ready()
+            logger.warning(f"Xorbits endpoint {self._external_web_endpoint} is ready!")
+            return self._external_web_endpoint
+        except:  # noqa: E722   # pragma: no cover
+            if self._log_when_fail:
+                logger.error("Error when creating cluster")
+                for name, log in self._load_cluster_logs().items():
+                    logger.error("Error logs for %s:\n%s", name, log)
+            self.stop()
+            raise
+
+    def stop(self, wait: bool = False, timeout: int = 0):
+        # stop isolation
+        stop_isolation()
+
+        from kubernetes.client import CoreV1Api
+
+        api = CoreV1Api(self._api_client)
+        api.delete_namespace(self._namespace)
+        if wait:
+            start_time = time.time()
+            while True:
+                try:
+                    api.read_namespace(self._namespace)
+                except K8SApiException as ex:  # pragma: no cover
+                    if ex.status != 404:
+                        raise
+                    break
+                else:
+                    time.sleep(1)
+                    if (
+                        timeout and time.time() - start_time > timeout
+                    ):  # pragma: no cover
+                        raise TimeoutError
+
+
+def new_cluster(
+    kube_api_client: Optional[ApiClient] = None,
+    image: Optional[str] = None,
+    supervisor_num: int = 1,
+    supervisor_cpu: Optional[int] = None,
+    supervisor_mem: Optional[str] = None,
+    worker_num: int = 1,
+    worker_cpu: Optional[int] = None,
+    worker_mem: Optional[str] = None,
+    worker_spill_paths: Optional[List[str]] = None,
+    worker_cache_mem: Optional[str] = None,
+    min_worker_num: Optional[int] = None,
+    web_num: int = 1,
+    web_cpu: Optional[int] = None,
+    web_mem: Optional[str] = None,
+    service_type: Optional[str] = None,
+    timeout: Optional[int] = None,
+    cluster_type: str = "minikube",
+    **kwargs,
+) -> "KubernetesClusterClient":
+    """
+    :param kube_api_client: Kubernetes API client, can be created with ``new_client_from_config``
```

Comment:
Consider to use `numpydoc` style which is mainly adopted in our docs to keep consistency, refer to https://stackoverflow.com/a/24385103

Reply:
Fix